File: scripts/check-migrations.mjs
#!/usr/bin/env node

// Pre-deployment migration checker
// Ensures migrations are applied before deploying new code

import fs from 'fs';
import path from 'path';
import { execSync } from 'child_process';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

console.log('🔍 Checking migration status...\n');

// Read all migration files
const migrationsDir = path.join(__dirname, '..', 'migrations');
const migrationFiles = fs.readdirSync(migrationsDir)
  .filter(f => f.endsWith('.sql'))
  .sort();

console.log(`Found ${migrationFiles.length} migration files:`);
migrationFiles.forEach(f => console.log(`  - ${f}`));

// Check which migrations have been applied
try {
  const databaseId = process.env.D1_DATABASE_ID;
  if (!databaseId) {
    console.error('\n❌ D1_DATABASE_ID environment variable not set');
    console.log('\nPlease set it with:');
    console.log('  export D1_DATABASE_ID=your-database-id\n');
    process.exit(1);
  }
  
  const result = execSync(`wrangler d1 migrations list ${databaseId}`, { encoding: 'utf8' });
  console.log('\n📋 Applied migrations:');
  console.log(result);

  // Parse applied migrations
  const appliedMigrations = result
    .split('\n')
    .filter(line => line.includes('.sql'))
    .map(line => line.trim().split(/\s+/)[0]);

  // Find unapplied migrations
  const unappliedMigrations = migrationFiles.filter(
    file => !appliedMigrations.some(applied => applied.includes(file))
  );

  if (unappliedMigrations.length > 0) {
    console.log('\n⚠️  Unapplied migrations found:');
    unappliedMigrations.forEach(f => console.log(`  - ${f}`));
    
    console.log('\n❌ Please apply migrations before deploying:');
    console.log('   wrangler d1 migrations apply omnichat-db\n');
    process.exit(1);
  } else {
    console.log('\n✅ All migrations are applied!');
  }

} catch (error) {
  console.error('\n❌ Error checking migrations:', error.message);
  console.log('\nMake sure you have:');
  console.log('  1. Configured wrangler with your Cloudflare credentials');
  console.log('  2. Created the omnichat-db database');
  process.exit(1);
}

File: scripts/check-migrations.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
)

func main() {
	fmt.Println("🔍 Checking migration status...\n")

	// Get database ID from environment
	databaseID := os.Getenv("D1_DATABASE_ID")
	if databaseID == "" {
		fmt.Println("❌ D1_DATABASE_ID environment variable not set")
		fmt.Println("\nPlease set it with:")
		fmt.Println("  export D1_DATABASE_ID=your-database-id\n")
		os.Exit(1)
	}

	// Read all migration files
	migrationsDir := filepath.Join("..", "migrations")
	files, err := ioutil.ReadDir(migrationsDir)
	if err != nil {
		log.Fatalf("Error reading migrations directory: %v", err)
	}

	var migrationFiles []string
	for _, file := range files {
		if strings.HasSuffix(file.Name(), ".sql") {
			migrationFiles = append(migrationFiles, file.Name())
		}
	}
	sort.Strings(migrationFiles)

	fmt.Printf("Found %d migration files:\n", len(migrationFiles))
	for _, f := range migrationFiles {
		fmt.Printf("  - %s\n", f)
	}

	// Check which migrations have been applied
	cmd := exec.Command("wrangler", "d1", "migrations", "list", databaseID)
	output, err := cmd.Output()
	if err != nil {
		fmt.Printf("\n❌ Error checking migrations: %v\n", err)
		fmt.Println("\nMake sure you have:")
		fmt.Println("  1. Configured wrangler with your Cloudflare credentials")
		fmt.Println("  2. Created the omnichat-db database")
		os.Exit(1)
	}

	fmt.Println("\n📋 Applied migrations:")
	fmt.Println(string(output))

	// Parse applied migrations
	lines := strings.Split(string(output), "\n")
	var appliedMigrations []string
	for _, line := range lines {
		if strings.Contains(line, ".sql") {
			fields := strings.Fields(line)
			if len(fields) > 0 {
				appliedMigrations = append(appliedMigrations, fields[0])
			}
		}
	}

	// Find unapplied migrations
	var unappliedMigrations []string
	for _, file := range migrationFiles {
		found := false
		for _, applied := range appliedMigrations {
			if strings.Contains(applied, file) {
				found = true
				break
			}
		}
		if !found {
			unappliedMigrations = append(unappliedMigrations, file)
		}
	}

	if len(unappliedMigrations) > 0 {
		fmt.Println("\n⚠️  Unapplied migrations found:")
		for _, f := range unappliedMigrations {
			fmt.Printf("  - %s\n", f)
		}
		fmt.Println("\n❌ Please apply migrations before deploying:")
		fmt.Printf("   wrangler d1 migrations apply %s\n\n", databaseID)
		os.Exit(1)
	}

	fmt.Println("\n✅ All migrations are applied!")
}